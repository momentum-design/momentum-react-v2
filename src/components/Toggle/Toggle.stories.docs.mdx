The `<Toggle />` a.k.a. switch input component

It looks similar to the checkbox input but it differs many ways:

- Toggle is usually used in the context where user expect immediate changes similar hwo lightswitch
  turn on/off light. For example: turn on/off settings.
- It has a bigger clickable surface which helps to mobile users

_Note:_ This component does not have label because it can appear in different layout setup, but it
must be wrapped within a `<label>` make it accessible.
