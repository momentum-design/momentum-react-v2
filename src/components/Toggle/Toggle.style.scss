@use "sass:math";

.md-toggle-switch {
  $switch-width: 2.75rem;
  $switch-height: math.div($switch-width, 2);
  $thumb-size: 0.635rem;
  $thumb-margin: math.div($switch-height - $thumb-size, 2);
  $transition-time: 150ms;

  appearance: none;
  width: $switch-width;
  height: $switch-height;
  border-radius: math.div($switch-height, 2);
  margin-right: 0.5rem;
  transition: background $transition-time;
  background-color: var(--mds-color-theme-control-inactive-normal);
  border: 1px solid var(--mds-color-theme-outline-input-normal);
  display: flex;
  justify-content: flex-start;
  align-items: center;

  &:before {
    content: '';
    width: $thumb-size;
    height: $thumb-size;
    border-radius: math.div($thumb-size, 2);
    margin-left: $thumb-margin;
    display: block;
    transition: transform $transition-time;
    background-color: var(--mds-color-theme-text-primary-normal);
  }

  &:checked:before {
    transform: translateX($switch-width - $switch-height);
  }

  &:checked {
    background-color: var(--mds-color-theme-control-active-normal);
    border: none;

    &:before {
      background-color: var(--mds-color-theme-inverted-text-primary-normal);
      transform: translateX($switch-width - $switch-height);
    }
  }

  // Focused
  &:focus {
    border: 1px solid var(--mds-color-theme-outline-input-normal);
    box-shadow: var(--md-globals-focus-ring-box-shadow);

    &:checked {
      border: none;
    }
  }

  // Hovered
  &:hover {
    background-color: var(--mds-color-theme-control-inactive-hover);

    &:before {
      background-color: var(--mds-color-theme-text-primary-normal);
    }

    &:checked {
      background-color: var(--mds-color-theme-control-active-hover);

      &:before {
        background-color: var(--mds-color-theme-inverted-text-primary-normal);
        transform: translateX($switch-width - $switch-height);
      }
    }
  }

  // Pressed
  &:active {
    background-color: var(--mds-color-theme-control-inactive-pressed);

    &:before {
      background-color: var(--mds-color-theme-inverted-text-primary-normal);
    }

    &:checked {
      background-color: var(--mds-color-theme-control-active-pressed);

      &:before {
        background-color: var(--mds-color-theme-inverted-text-primary-normal);
        transform: translateX($switch-width - $switch-height);
      }
    }
  }

  // Disabled
  &:disabled {
    background-color: var(--mds-color-theme-control-inactive-disabled);

    &:before {
      background-color: var(--mds-color-theme-text-primary-disabled);
    }

    &:checked {
      background-color: var(--mds-color-theme-control-active-disabled);

      &:before {
        background-color: var(--mds-color-theme-inverted-text-primary-disabled);
        transform: translateX($switch-width - $switch-height);
      }
    }
  }
}
