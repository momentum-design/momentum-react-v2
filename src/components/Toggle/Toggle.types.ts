import { CSSProperties } from 'react';
import { AriaSwitchBase } from '@react-types/switch';
import { AriaLabelRequired } from '../../utils/a11y';

export type Props = Omit<AriaSwitchBase, 'children'> & {
  /**
   * Custom class for overriding this component's CSS.
   */
  className?: string;

  /**
   * Custom id for overriding this component's CSS.
   */
  id?: string;

  /**
   * Custom style for overriding this component's CSS.
   */
  style?: CSSProperties;
} & AriaLabelRequired;
