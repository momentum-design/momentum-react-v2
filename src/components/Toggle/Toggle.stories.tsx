import { MultiTemplate, Template } from '../../storybook/helper.stories.templates';
import { DocumentationPage } from '../../storybook/helper.stories.docs';
import StyleDocs from '../../storybook/docs.stories.style.mdx';

import Toggle, { ToggleProps, TOGGLE_CONSTANTS } from './';
import Documentation from './Toggle.stories.docs.mdx';
import argTypes from './Toggle.stories.args';
import React from 'react';

export default {
  title: 'Momentum UI/Toggle',
  component: Toggle,
  parameters: {
    expanded: true,
    docs: {
      page: DocumentationPage(Documentation, StyleDocs),
    },
  },
};

type ToggleWithLabelProps = ToggleProps & { label: string };
const ToggleWithLabel = ({ label, ...props }: ToggleWithLabelProps) => (
  <div>
    <label htmlFor={label} style={{ display: 'flex', alignItems: 'center' }}>
      <Toggle id={label} {...props} />
      {label}
    </label>
  </div>
);

const Example = Template<ToggleProps>(Toggle);

Example.argTypes = { ...argTypes };

Example.args = {
  defaultSelected: TOGGLE_CONSTANTS.DEFAULTS.DEFAULT_SELECTION,
  isDisabled: TOGGLE_CONSTANTS.DEFAULTS.IS_DISABLED,
};

const Common = MultiTemplate<ToggleWithLabelProps>(ToggleWithLabel);

Common.argTypes = { ...argTypes };
delete Common.argTypes.defaultSelected;
delete Common.argTypes.isDisabled;
delete Common.argTypes.label;
delete Common.argTypes.onChange;

Common.args = {};
Common.parameters = {
  variants: [
    { defaultSelected: true, isDisabled: false, label: 'Selected' },
    { defaultSelected: true, isDisabled: true, label: 'Selected + Disabled' },
    { defaultSelected: false, isDisabled: false, label: 'Not selected' },
    { defaultSelected: false, isDisabled: true, label: 'Not selected + Disabled' },
  ],
};

export { Example, Common };
