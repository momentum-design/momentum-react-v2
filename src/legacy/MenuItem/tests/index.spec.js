import React from 'react';
import { mount } from 'enzyme';
import SelectableContext from '../../SelectableContext';
import { Icon, ListItemSection, MenuItemLegacy } from '@momentum-ui/react-collaboration';
import Toggle from '../../../components/Toggle';
beforeEach(() => {
  jest.resetModules();
  jest.clearAllMocks();
});

describe('tests for <MenuItemLegacy />', () => {
  const context = {
    parentKeyDown: jest.fn(),
    parentKeyUp: jest.fn(),
    parentOnSelect: jest.fn(),
  };

  it('should render a MenuItem', () => {
    const wrapper = mount(<MenuItemLegacy label="one" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should call handleClick function of context when ListItem is clicked', () => {
    const onClick = jest.fn();

    const wrapper = mount(
      <SelectableContext.Provider value={context}>
        <MenuItemLegacy onClick={onClick} label="one" />
      </SelectableContext.Provider>
    );

    const listItem = wrapper.find('ListItem').first();
    listItem.simulate('click');
    expect(onClick).toHaveBeenCalled();
    expect(context.parentOnSelect.mock.calls.length).toBe(1);
  });

  it('should pass value to onClick handler', () => {
    const onClick = jest.fn();

    const wrapper = mount(
      <SelectableContext.Provider value={context}>
        <MenuItemLegacy onClick={onClick} label="one" value="test" />
      </SelectableContext.Provider>
    );

    const listItem = wrapper.find('ListItem').first();
    listItem.simulate('click');
    expect(context.parentOnSelect.mock.calls.length).toBe(1);
    expect(onClick.mock.calls[0][1].value).toBe('test');
  });

  it('should not call e.preventDefault if target is not md-list-item', () => {
    const onClick = jest.fn();

    const wrapper = mount(
      <SelectableContext.Provider value={context}>
        <MenuItemLegacy onClick={onClick} label="one" value="test">
          <ListItemSection position="left">
            <Icon name="edit_20" />
          </ListItemSection>
          <ListItemSection position="center">Dark mode</ListItemSection>
          <ListItemSection position="right">
            <Toggle aria-label="toggle" />
          </ListItemSection>
        </MenuItemLegacy>
      </SelectableContext.Provider>
    );

    const toggle = wrapper.find(Toggle);

    const mockEvent = {
      preventDefault: jest.fn(),
      keyCode: 32,
      target: toggle.getDOMNode(),
    };

    const listItem = wrapper.find(Toggle);
    listItem.simulate('keydown', mockEvent);
    expect(onClick).not.toBeCalled();
    expect(context.parentKeyDown).not.toBeCalled();
    expect(mockEvent.preventDefault).not.toBeCalled();
  });

  it('should call handleKeyDown function of context when keyDown is fired on ListItem', () => {
    const onKeyDown = jest.fn();

    const wrapper = mount(
      <SelectableContext.Provider value={context}>
        <MenuItemLegacy onKeyDown={onKeyDown} label="one" value="test" />
      </SelectableContext.Provider>
    );

    const listItem = wrapper.find('ListItem').first();
    listItem.simulate('keydown');
    expect(context.parentKeyDown.mock.calls.length).toBe(1);
    expect(onKeyDown).toHaveBeenCalled();
  });

  it('should call handleKeyUp function of context when keyUp is fired on ListItem', () => {
    const onKeyUp = jest.fn();

    const wrapper = mount(
      <SelectableContext.Provider value={context}>
        <MenuItemLegacy onKeyUp={onKeyUp} label="one" value="test" />
      </SelectableContext.Provider>
    );

    const listItem = wrapper.find('ListItem').first();
    listItem.simulate('keyup');
    expect(context.parentKeyUp.mock.calls.length).toBe(1);
    expect(onKeyUp).toHaveBeenCalled();
  });

  it('should apply class for isHeader prop', () => {
    const container = mount(<MenuItemLegacy isHeader />);

    expect(container.find('.md-menu-item__header').exists()).toEqual(true);
  });

  describe('tests for title Prop', () => {
    it('should not have title by default', () => {
      const container = mount(<MenuItemLegacy />);

      expect(container.find('.md-list-item').props().title).toEqual(undefined);
    });

    it('should handle title prop', () => {
      const container = mount(<MenuItemLegacy title="testTitle" />);

      expect(container.find('.md-list-item').props().title).toEqual('testTitle');
    });
  });
});
