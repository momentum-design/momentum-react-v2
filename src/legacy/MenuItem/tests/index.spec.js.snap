// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for <MenuItem /> should render a MenuItem 1`] = `
<ContextTransform(MenuItem)
  label="one"
>
  <MenuItem
    className=""
    isHeader={false}
    isOpen={false}
    itemClassName=""
    keepMenuOpen={false}
    label="one"
    onClick={null}
    parentKeyDown={null}
    parentOnSelect={null}
  >
    <Component
      name={[Function]}
    >
      <UID
        idSource={
          {
            "prefix": "",
            "uid": [Function],
            "value": 3,
          }
        }
        name={[Function]}
      >
        <div
          aria-expanded={null}
          aria-haspopup={false}
          className="md-menu-item"
        >
          <ContextTransform(ListItem)
            active={null}
            className=""
            focusOnLoad={true}
            id="md-menu-item-1"
            isReadOnly={false}
            label="one"
            role="menuitem"
          >
            <ListItem
              active={null}
              className=""
              customAnchorNode={null}
              customRefProp=""
              disabled={false}
              eventKey=""
              focus={false}
              focusLockTabbableChildren={false}
              focusLockTabbableChildrenProps={
                {
                  "portalNodeQuery": "",
                  "tabbableChildrenHasPopover": false,
                  "tabbableChildrenQuery": "",
                }
              }
              focusOnLoad={true}
              id="md-menu-item-1"
              isReadOnly={false}
              itemIndex={null}
              keyboardKey=""
              label="one"
              link=""
              onBlur={null}
              onClick={null}
              onKeyDown={null}
              parentKeyDown={[Function]}
              parentOnSelect={[Function]}
              refName="navLink"
              role="menuitem"
              separator={false}
              specifyRoleWithoutList={false}
              title=""
              type=""
              value=""
            >
              <Component
                name={[Function]}
              >
                <UID
                  idSource={
                    {
                      "prefix": "",
                      "uid": [Function],
                      "value": 3,
                    }
                  }
                  name={[Function]}
                >
                  <div
                    className="md-list-item"
                    data-md-event-key="md-menu-item-1"
                    data-md-keyboard-key="one"
                    id="md-menu-item-1"
                    onBlur={[Function]}
                    onClick={[Function]}
                    onKeyDown={[Function]}
                    role="menuitem"
                    tabIndex={-1}
                    title="one"
                  >
                    one
                  </div>
                </UID>
              </Component>
            </ListItem>
          </ContextTransform(ListItem)>
        </div>
      </UID>
    </Component>
  </MenuItem>
</ContextTransform(MenuItem)>
`;
